runtime/gc_vlhgc/CopyForwardScheme.cpp

```diff
@@ -1991,24 +1990,9 @@ MM_CopyForwardScheme::copy(MM_EnvironmentVLHGC *env, MM_AllocationContextTarok *
 
 				if (objectModel->isIndexable(destinationObjectPtr)) {
 					_extensions->indexableObjectModel.fixupInternalLeafPointersAfterCopy((J9IndexableObject *)destinationObjectPtr, (J9IndexableObject *)forwardedHeader->getObject());
-
-					/* Updates internal data address of indexable objects. Every indexable object have a void *dataAddr
-					 * that always points to the array data. It will always point to the address right after the header,
-					 * in case of contiguous data it will point to the data itself, and in case of discontiguous
-					 * arraylet it will point to the first arrayiod. dataAddr is only updated if dataAddr points to data
-					 * within heap. */
-					_extensions->indexableObjectModel.fixupDataAddr(destinationObjectPtr);
 				}
 
-				/* IF the object has been hashed and has not been moved then we must store the previous
-				 * address into the hashcode slot at hashcode offset.
-				 */
-				if (doesObjectNeedHash) {
-					UDATA hashOffset = objectModel->getHashcodeOffset(destinationObjectPtr);
-					U_32 *hashCodePointer = (U_32*)((U_8*)destinationObjectPtr + hashOffset);
-					*hashCodePointer = objectModel->computeObjectHash(forwardedHeader);
-					objectModel->setObjectHasBeenMoved(destinationObjectPtr);
-				}
+				_extensions->objectModel.fixupForwardedObject(forwardedHeader, destinationObjectPtr, _extensions->objectModel.getPreservedAge(forwardedHeader));
```

Comment:
I like this solution. As expected this will result in some code duplication in ObjectModel, but I presume this tradeoff has already been considered?

Reply:
Amount of executable code duplication is not an issue. Amount of source code duplication might be an issue (from code maintenance point of view), depending on a win %, but it'll certainly be low, if measurable at all.

We can consider a compromise/hybrid and share hashing code block (via a separate inline method like fixupHashInfo), which would 
- either take hashSlot as an argument prepared by caller fixupForwardedObjectMixed/Indexable (probably not good because it would be calculated even in cases it's not needed, which indeed is the case),
- or keep isIndexable check in it (probably better since, it would be infrequently exercised - only on first move of a hashed object)

Reply:
I'm also realizing that having fixupForwardedObjectMixed/Indexable inside common ObjectModel class is not a clean solution. It would be cleaner to have fixupForwardedObject in already specialized classes Mixed/ArrayObjectModel, but than it may impose some other difficulties... :)

Reply:
I see the complications for sure. I implemented the fixupForwardedObjectMixed/Indexable but as you mentioned the solution is not a particularly clean solution. Perhaps leaving as-in is the best for now?